Add installation guide to README

hanchenye opened the pull request:
As the llvm-project has became a submodule, I assume this is the correct building steps?

Got errors on the Polybench test cases:
```
mlir-clang: ../lib/polygeist/Ops.cpp:247: virtual mlir::LogicalResult DeallocSubView::matchAndRewrite(mlir::polygeist::SubIndexOp, mlir::PatternRewriter&) const: Assertion `subindex.source() .getType() .cast<MemRefType>() .getShape() .size() == indices.size()' failed.
```
It seems `mlir-clang %s %polyverify %stdinclude -emit-llvm` has error out.

Did I miss anything in the installation?

chelini commented:
> As the llvm-project has became a submodule, I assume this is the correct building steps?
> 
> Got errors on the Polybench test cases:
> 
> ```
> mlir-clang: ../lib/polygeist/Ops.cpp:247: virtual mlir::LogicalResult DeallocSubView::matchAndRewrite(mlir::polygeist::SubIndexOp, mlir::PatternRewriter&) const: Assertion `subindex.source() .getType() .cast<MemRefType>() .getShape() .size() == indices.size()' failed.
> ```
> 
> It seems `mlir-clang %s %polyverify %stdinclude -emit-llvm` has error out.
> 
> Did I miss anything in the installation?

You need to run `ninja` before `check-mlir-clang`, the reason is that check-mlir-clang does not build polygeist at the moment. We need to fix the build infrastructure.

chelini commented on README.md:
```diff
@@ -1,106 +1,35 @@
+mkdir $SOURCE_DIR/llvm-project/build
+cd $SOURCE_DIR/llvm-project/build
+cmake -G Ninja ../llvm \
+  -DLLVM_ENABLE_PROJECTS="mlir;llvm;clang" \
```
llvm is not needed.

chelini commented on README.md:
```diff
@@ -1,106 +1,35 @@
+cd $SOURCE_DIR/build
+cmake -G Ninja .. \
+  -DMLIR_DIR=$PWD/../llvm/build/lib/cmake/mlir \
+  -DLLVM_DIR=$PWD/../llvm/build/lib/cmake/llvm \
```
same here.

chelini commented on README.md:
```diff
@@ -1,106 +1,35 @@
+  -DCLANG_DIR=$PWD/../llvm/build/lib/cmake/clang \
+  -DLLVM_ENABLE_ASSERTIONS=ON \
+  -DCMAKE_BUILD_TYPE=DEBUG
+ninja check-mlir-clang
```
ninja before ninja check-mlir-clang. But of course, a better fix would be welcome.

chelini commented on README.md:
```diff
@@ -1,106 +1,35 @@
-# The LLVM Compiler Infrastructure
```
Last time we moved the build instruction in the mlir-clang directory, and we kept this README. I would prefer to keep this file and move these changes in the mlir-clang dir.

chelini commented on README.md:
```diff
@@ -1,106 +1,35 @@
+mkdir $SOURCE_DIR/build
+cd $SOURCE_DIR/build
+cmake -G Ninja .. \
+  -DMLIR_DIR=$PWD/../llvm/build/lib/cmake/mlir \
```
here I thinks it should be llvm-project and not llvm?

chelini commented:
@hanchenye did you manage to solve the problem?

hanchenye commented:
> I am able to reproduce your error by manually running the command, check-mlir-clang on my machine works, without any issues. The error is related to -D POLYBENCH_TIME

@chelini Right, `mlir-clang` is doing well on my machine, too. The errors are still there with `ninja` executed before `check-mlir-clang`.

Updated the instruction and moved it to `mlir-clang` directory.

wsmoses commented:
Pardon the delay was trying to make an aggressive paper deadline last week and didn't have a chance to investigate. I've reproduced the issue locally and will hopefully fix shortly now.

wsmoses commented:
@chelini why do you prefer having install instructions within mlir-clang. The current README shouldn't actually be here since it's a leftover from LLVM (which has been removed). I have a slight preference for having the instructions at the top since then the top level read me contain instructions for using the code in this repo.

wsmoses commented on mlir-clang/README.md:
````diff
@@ -1,36 +1,42 @@
-             to use ninja instead of make make # or ninja; preferably with -j <num-cores>
+### 0. Clone Polygeist
+```sh
+git clone git@github.com:wsmoses/Polygeist.git $SOURCE_DIR
````
I'd change this to be a clone via https. Also you can remote the submodule init/update if you make the clone recursive via `--recursive`.

wsmoses commented on mlir-clang/README.md:
```diff
@@ -1,36 +1,42 @@
+cmake -G Ninja .. \
+  -DMLIR_DIR=$PWD/../llvm-project/build/lib/cmake/mlir \
+  -DCLANG_DIR=$PWD/../llvm-project/build/lib/cmake/clang \
+  -DLLVM_ENABLE_ASSERTIONS=ON \
```
I would remove the assertion/build type here, and also add the LLVM_EXTERNAL_LIT flag as here: https://github.com/wsmoses/Polygeist/blob/8fb9ac6429957b4c9206a775b2bb9c06e2799e18/.github/workflows/build.yml#L56

stephenneuendorffer replied:
LLVM_EXTERNAL_LIT should have a proper default if you're compiling against a build directory.  It should only be necessary if you're compiling against an install directory.

wsmoses replied:
Didn't know that! Yeah in which case so long as we're always building LLVM from scratch against the custom hash, no need for LLVM_EXTERNAL_LIT

hanchenye commented:
@wsmoses I've updated the PR. Agree with you that the original LLVM README should retire.

wsmoses commented on README.md:
````diff
@@ -1,106 +1,40 @@
+ninja check-mlir-clang
+```
+
+Note that everything is statically linked, the produced binary should be freely movable around the file system.
````
I'd refrain from making the static link comment.

wsmoses commented on README.md:
```diff
@@ -1,106 +1,40 @@
+  -DLLVM_ENABLE_ASSERTIONS=ON \
+  -DCMAKE_BUILD_TYPE=DEBUG
+ninja
+ninja check-mlir-clang
```
Could you add an example?

If possible, a couple of other things to add:
* You will almost certainly want to add the resource directory of the clang built here to the command line `-resource-dir ~/git/Polygeist/mlir-build/lib/clang/14.0.0/`
* Defines, libraries, includes, etc are supported but require space separation.

Perhaps a full example with a 2 line c code?

hanchenye replied:
I'd suggest to add a more comprehensive command line tool reference for this? Something like the [command line guide of clang](https://github.com/llvm/llvm-project/blob/c2574e63ff71c1d3caea48cb6200c2422bd8f33d/clang/docs/CommandGuide/clang.rst).

wsmoses replied:
Fair enough -- and for that (and other things) we should actually put content in the placeholder website branch.

hanchenye replied:
Makes sense to me!

wsmoses reviewed (approved):
LGTM, but would recommend a couple of README updates.

wsmoses commented on README.md:
````diff
@@ -1,106 +1,40 @@
+
+### 0. Clone Polygeist
+```sh
+git clone --recursive https://github.com/wsmoses/Polygeist.git $SOURCE_DIR
````
Might remove SOURCE_DIR as a variable and just use Polygeist for now to simplify.

chelini commented:
Sorry for the latency I took a week off. Indeed now the readme should not go in mlir-clang. I was still thinking for some reason about the old non-standalone frontend. Sorry for the confusion @hanchenye.

mikeurbach commented:
After this lands, assuming #65 is eventually accepted, I can include a section about installation using LLVM_EXTERNAL_PROJECTS.
